File: packages/core/tests/classify.test.ts
/**
 * Contract tests para `classifyField`.
 *
 * Estos tests definen el **corpus canónico** de inputs (combinaciones
 * field+behavior) con sus outputs esperados. Cualquier SDK mirror
 * (futuro `protocol-dart` cuando exista Flutter) debe producir el mismo
 * output para los mismos inputs — garantía de paridad cross-language
 * por construcción.
 *
 * Si añades un behavior nuevo con `renderAsSlotKind`, también añade un
 * caso aquí cubriendo:
 *  - cómo se clasifica el field cuando tiene ese behavior
 *  - cómo se clasifica cuando tiene el `type` base pero NO el behavior
 *  - cualquier excepción semántica (ej: color_hex es type=text pero NO
 *    cae en text-short)
 *
 * El corpus también es ground truth para el visualizador del grafo
 * (KRO-71 Fase 3) y para los tooltips de la enciclopedia (KRO-70).
 */

import { describe, it, expect } from 'vitest';
import { classifyField, isFieldCompatibleWithSlot } from '../src/classify';
import type { SlotAcceptKind } from '../src/types';
import type { SlotDefinition } from '../src/registries/recipes';

/** Una entry del corpus: input + outputs esperados (orden ignorado). */
interface CorpusEntry {
  /** Descripción legible para el `it()`. */
  desc: string;
  field: { type: string; behavior?: string };
  /** Kinds que `classifyField` debe devolver. 'any' siempre incluido. */
  expectedKinds: SlotAcceptKind[];
}

const CORPUS: CorpusEntry[] = [
  // ── type=image ────────────────────────────────────────────────────
  {
    desc: 'image sin behavior → image + 3 legacy aliases',
    field: { type: 'image' },
    expectedKinds: ['any', 'image', 'image-avatar', 'image-banner', 'image-cover'],
  },

  // ── type=array<image> ─────────────────────────────────────────────
  {
    desc: 'array<image> sin behavior → image-array',
    field: { type: 'array<image>' },
    expectedKinds: ['any', 'image-array'],
  },
  {
    desc: 'array<image> con behavior=gallery → image-array',
    field: { type: 'array<image>', behavior: 'gallery' },
    expectedKinds: ['any', 'image-array'],
  },
  {
    desc: 'array<image> con behavior=slideshow → image-array',
    field: { type: 'array<image>', behavior: 'slideshow' },
    expectedKinds: ['any', 'image-array'],
  },
  {
    desc: 'array<image> con behavior=card_multiview → image-array',
    field: { type: 'array<image>', behavior: 'card_multiview' },
    expectedKinds: ['any', 'image-array'],
  },

  // ── type=number ───────────────────────────────────────────────────
  {
    desc: 'number sin behavior → number + text-short',
    field: { type: 'number' },
    expectedKinds: ['any', 'number', 'text-short'],
  },
  {
    desc: 'number con behavior=year → date + number + text-short',
    field: { type: 'number', behavior: 'year' },
    expectedKinds: ['any', 'date', 'number', 'text-short'],
  },
  {
    desc: 'number con behavior=rating → badge + number + text-short',
    field: { type: 'number', behavior: 'rating' },
    expectedKinds: ['any', 'badge', 'number', 'text-short'],
  },
  {
    desc: 'number con behavior=currency → solo number + text-short (currency no tiene renderAsSlotKind)',
    field: { type: 'number', behavior: 'currency' },
    expectedKinds: ['any', 'number', 'text-short'],
  },
  {
    desc: 'number con behavior=percentage → solo number + text-short',
    field: { type: 'number', behavior: 'percentage' },
    expectedKinds: ['any', 'number', 'text-short'],
  },

  // ── type=text ─────────────────────────────────────────────────────
  {
    desc: 'text sin behavior → text-short',
    field: { type: 'text' },
    expectedKinds: ['any', 'text-short'],
  },
  {
    desc: 'text con behavior=url → url + text-short',
    field: { type: 'text', behavior: 'url' },
    expectedKinds: ['any', 'url', 'text-short'],
  },
  {
    desc: 'text con behavior=email → url + text-short',
    field: { type: 'text', behavior: 'email' },
    expectedKinds: ['any', 'url', 'text-short'],
  },
  {
    desc: 'text con behavior=phone → url + text-short',
    field: { type: 'text', behavior: 'phone' },
    expectedKinds: ['any', 'url', 'text-short'],
  },
  {
    desc: 'text con behavior=iso_date → date + text-short',
    field: { type: 'text', behavior: 'iso_date' },
    expectedKinds: ['any', 'date', 'text-short'],
  },
  {
    desc: 'text con behavior=color_hex → color (NO text-short, excepción KRO-69)',
    field: { type: 'text', behavior: 'color_hex' },
    expectedKinds: ['any', 'color'],
  },
  {
    desc: 'text con behavior=ordinal_enum → badge + text-short',
    field: { type: 'text', behavior: 'ordinal_enum' },
    expectedKinds: ['any', 'badge', 'text-short'],
  },
  {
    desc: 'text con behavior=slug → solo text-short (slug no tiene renderAsSlotKind)',
    field: { type: 'text', behavior: 'slug' },
    expectedKinds: ['any', 'text-short'],
  },

  // ── type=textarea ─────────────────────────────────────────────────
  {
    desc: 'textarea sin behavior → text-long',
    field: { type: 'textarea' },
    expectedKinds: ['any', 'text-long'],
  },
  {
    desc: 'textarea con behavior=markdown → text-long',
    field: { type: 'textarea', behavior: 'markdown' },
    expectedKinds: ['any', 'text-long'],
  },
  {
    desc: 'textarea con behavior=notes → text-long',
    field: { type: 'textarea', behavior: 'notes' },
    expectedKinds: ['any', 'text-long'],
  },
  {
    desc: 'textarea con behavior=html → text-long',
    field: { type: 'textarea', behavior: 'html' },
    expectedKinds: ['any', 'text-long'],
  },
  {
    desc: 'textarea con behavior=code → solo text-long (code no tiene renderAsSlotKind)',
    field: { type: 'textarea', behavior: 'code' },
    expectedKinds: ['any', 'text-long'],
  },

  // ── type=select ───────────────────────────────────────────────────
  {
    desc: 'select sin behavior → text-short',
    field: { type: 'select' },
    expectedKinds: ['any', 'text-short'],
  },
  {
    desc: 'select con behavior=ordinal_enum → badge + text-short',
    field: { type: 'select', behavior: 'ordinal_enum' },
    expectedKinds: ['any', 'badge', 'text-short'],
  },

  // ── type=array<string> ─────────────────────────────────────────────
  {
    desc: 'array<string> con behavior=enum → badge',
    field: { type: 'array<string>', behavior: 'enum' },
    expectedKinds: ['any', 'badge'],
  },
  {
    desc: 'array<string> con behavior=card_code_list → card-ref',
    field: { type: 'array<string>', behavior: 'card_code_list' },
    expectedKinds: ['any', 'card-ref'],
  },
  {
    desc: 'array<string> con behavior=tags → solo any (tags no tiene renderAsSlotKind)',
    field: { type: 'array<string>', behavior: 'tags' },
    expectedKinds: ['any'],
  },

  // ── type=array<number> ─────────────────────────────────────────────
  {
    desc: 'array<number> con behavior=card_index_list → card-ref',
    field: { type: 'array<number>', behavior: 'card_index_list' },
    expectedKinds: ['any', 'card-ref'],
  },
  {
    desc: 'array<number> con behavior=year_list → date',
    field: { type: 'array<number>', behavior: 'year_list' },
    expectedKinds: ['any', 'date'],
  },

  // ── Edge cases ─────────────────────────────────────────────────────
  {
    desc: 'type desconocido → solo any',
    field: { type: 'unknown_future_type' },
    expectedKinds: ['any'],
  },
  {
    desc: 'behavior desconocido sobre text → text-short (fallback al type)',
    field: { type: 'text', behavior: 'mystery_behavior' },
    expectedKinds: ['any', 'text-short'],
  },
];

describe('classifyField — corpus canónico (cross-language ground truth)', () => {
  for (const entry of CORPUS) {
    it(entry.desc, () => {
      const result = classifyField(entry.field);
      // Ordenamos ambos lados para que el test no dependa del orden interno.
      expect([...result].sort()).toEqual([...entry.expectedKinds].sort());
    });
  }

  it('siempre incluye "any"', () => {
    for (const entry of CORPUS) {
      const kinds = classifyField(entry.field);
      expect(kinds).toContain('any');
    }
  });

  it('no devuelve duplicados', () => {
    for (const entry of CORPUS) {
      const kinds = classifyField(entry.field);
      const unique = new Set(kinds);
      expect(unique.size).toBe(kinds.length);
    }
  });
});

describe('isFieldCompatibleWithSlot', () => {
  const mkSlot = (accepts: SlotAcceptKind[]): SlotDefinition => ({
    id:      'test',
    label:   'Test',
    kind:    'single',
    accepts,
  });

  it('slot any-accepts admite cualquier field', () => {
    const slot = mkSlot(['any']);
    expect(isFieldCompatibleWithSlot({ type: 'text' }, slot)).toBe(true);
    expect(isFieldCompatibleWithSlot({ type: 'image' }, slot)).toBe(true);
    expect(isFieldCompatibleWithSlot({ type: 'unknown' }, slot)).toBe(true);
  });

  it('slot image admite type=image', () => {
    const slot = mkSlot(['image']);
    expect(isFieldCompatibleWithSlot({ type: 'image' }, slot)).toBe(true);
    expect(isFieldCompatibleWithSlot({ type: 'text' }, slot)).toBe(false);
  });

  it('slot text-short NO admite color_hex (KRO-69 exception)', () => {
    const slot = mkSlot(['text-short']);
    expect(isFieldCompatibleWithSlot({ type: 'text', behavior: 'color_hex' }, slot)).toBe(false);
  });

  it('slot text-short admite number (number cae en text-short)', () => {
    const slot = mkSlot(['text-short']);
    expect(isFieldCompatibleWithSlot({ type: 'number' }, slot)).toBe(true);
  });

  it('slot color SOLO admite behavior=color_hex', () => {
    const slot = mkSlot(['color']);
    expect(isFieldCompatibleWithSlot({ type: 'text', behavior: 'color_hex' }, slot)).toBe(true);
    expect(isFieldCompatibleWithSlot({ type: 'text' }, slot)).toBe(false);
    expect(isFieldCompatibleWithSlot({ type: 'image' }, slot)).toBe(false);
  });

  it('slot badge admite rating/enum/ordinal_enum', () => {
    const slot = mkSlot(['badge']);
    expect(isFieldCompatibleWithSlot({ type: 'number', behavior: 'rating' }, slot)).toBe(true);
    expect(isFieldCompatibleWithSlot({ type: 'array<string>', behavior: 'enum' }, slot)).toBe(true);
    expect(isFieldCompatibleWithSlot({ type: 'text', behavior: 'ordinal_enum' }, slot)).toBe(true);
    expect(isFieldCompatibleWithSlot({ type: 'text' }, slot)).toBe(false);
  });
});

File: packages/core_dart/test/classify_test.dart
/// Corpus canonico de `classifyField` — ESPEJO 1:1 de
/// `packages/core/tests/classify.test.ts`.
///
/// Mismos inputs → mismos outputs que el TS. Esta es la garantia de paridad
/// cross-language POR CONSTRUCCION: si el Dart diverge del TS en cualquier
/// caso, este test falla. Al añadir un caso al corpus TS, añadelo aqui igual.
import 'package:test/test.dart';
import 'package:kromia_core/kromia_core.dart';

class CorpusEntry {
  final String desc;
  final FieldSpec field;
  final List<String> expectedKinds;
  const CorpusEntry(this.desc, this.field, this.expectedKinds);
}

final List<CorpusEntry> corpus = <CorpusEntry>[
  // ── type=image ────────────────────────────────────────────────────
  CorpusEntry('image sin behavior -> image + 3 legacy aliases',
      const FieldSpec('image'),
      const ['any', 'image', 'image-avatar', 'image-banner', 'image-cover']),

  // ── type=array<image> ─────────────────────────────────────────────
  CorpusEntry('array<image> sin behavior -> image-array',
      const FieldSpec('array<image>'), const ['any', 'image-array']),
  CorpusEntry('array<image> con behavior=gallery -> image-array',
      const FieldSpec('array<image>', behavior: 'gallery'), const ['any', 'image-array']),
  CorpusEntry('array<image> con behavior=slideshow -> image-array',
      const FieldSpec('array<image>', behavior: 'slideshow'), const ['any', 'image-array']),
  CorpusEntry('array<image> con behavior=card_multiview -> image-array',
      const FieldSpec('array<image>', behavior: 'card_multiview'), const ['any', 'image-array']),

  // ── type=number ───────────────────────────────────────────────────
  CorpusEntry('number sin behavior -> number + text-short',
      const FieldSpec('number'), const ['any', 'number', 'text-short']),
  CorpusEntry('number con behavior=year -> date + number + text-short',
      const FieldSpec('number', behavior: 'year'), const ['any', 'date', 'number', 'text-short']),
  CorpusEntry('number con behavior=rating -> badge + number + text-short',
      const FieldSpec('number', behavior: 'rating'), const ['any', 'badge', 'number', 'text-short']),
  CorpusEntry('number con behavior=currency -> solo number + text-short',
      const FieldSpec('number', behavior: 'currency'), const ['any', 'number', 'text-short']),
  CorpusEntry('number con behavior=percentage -> solo number + text-short',
      const FieldSpec('number', behavior: 'percentage'), const ['any', 'number', 'text-short']),

  // ── type=text ─────────────────────────────────────────────────────
  CorpusEntry('text sin behavior -> text-short',
      const FieldSpec('text'), const ['any', 'text-short']),
  CorpusEntry('text con behavior=url -> url + text-short',
      const FieldSpec('text', behavior: 'url'), const ['any', 'url', 'text-short']),
  CorpusEntry('text con behavior=email -> url + text-short',
      const FieldSpec('text', behavior: 'email'), const ['any', 'url', 'text-short']),
  CorpusEntry('text con behavior=phone -> url + text-short',
      const FieldSpec('text', behavior: 'phone'), const ['any', 'url', 'text-short']),
  CorpusEntry('text con behavior=iso_date -> date + text-short',
      const FieldSpec('text', behavior: 'iso_date'), const ['any', 'date', 'text-short']),
  CorpusEntry('text con behavior=color_hex -> color (NO text-short, KRO-69)',
      const FieldSpec('text', behavior: 'color_hex'), const ['any', 'color']),
  CorpusEntry('text con behavior=ordinal_enum -> badge + text-short',
      const FieldSpec('text', behavior: 'ordinal_enum'), const ['any', 'badge', 'text-short']),
  CorpusEntry('text con behavior=slug -> solo text-short',
      const FieldSpec('text', behavior: 'slug'), const ['any', 'text-short']),

  // ── type=textarea ─────────────────────────────────────────────────
  CorpusEntry('textarea sin behavior -> text-long',
      const FieldSpec('textarea'), const ['any', 'text-long']),
  CorpusEntry('textarea con behavior=markdown -> text-long',
      const FieldSpec('textarea', behavior: 'markdown'), const ['any', 'text-long']),
  CorpusEntry('textarea con behavior=notes -> text-long',
      const FieldSpec('textarea', behavior: 'notes'), const ['any', 'text-long']),
  CorpusEntry('textarea con behavior=html -> text-long',
      const FieldSpec('textarea', behavior: 'html'), const ['any', 'text-long']),
  CorpusEntry('textarea con behavior=code -> solo text-long',
      const FieldSpec('textarea', behavior: 'code'), const ['any', 'text-long']),

  // ── type=select ───────────────────────────────────────────────────
  CorpusEntry('select sin behavior -> text-short',
      const FieldSpec('select'), const ['any', 'text-short']),
  CorpusEntry('select con behavior=ordinal_enum -> badge + text-short',
      const FieldSpec('select', behavior: 'ordinal_enum'), const ['any', 'badge', 'text-short']),

  // ── type=array<string> ────────────────────────────────────────────
  CorpusEntry('array<string> con behavior=enum -> badge',
      const FieldSpec('array<string>', behavior: 'enum'), const ['any', 'badge']),
  CorpusEntry('array<string> con behavior=card_code_list -> card-ref',
      const FieldSpec('array<string>', behavior: 'card_code_list'), const ['any', 'card-ref']),
  CorpusEntry('array<string> con behavior=tags -> solo any',
      const FieldSpec('array<string>', behavior: 'tags'), const ['any']),

  // ── type=array<number> ────────────────────────────────────────────
  CorpusEntry('array<number> con behavior=card_index_list -> card-ref',
      const FieldSpec('array<number>', behavior: 'card_index_list'), const ['any', 'card-ref']),
  CorpusEntry('array<number> con behavior=year_list -> date',
      const FieldSpec('array<number>', behavior: 'year_list'), const ['any', 'date']),

  // ── Edge cases ────────────────────────────────────────────────────
  CorpusEntry('type desconocido -> solo any',
      const FieldSpec('unknown_future_type'), const ['any']),
  CorpusEntry('behavior desconocido sobre text -> text-short (fallback al type)',
      const FieldSpec('text', behavior: 'mystery_behavior'), const ['any', 'text-short']),
];

void main() {
  group('classifyField — corpus canonico (cross-language ground truth)', () {
    for (final entry in corpus) {
      test(entry.desc, () {
        final result = classifyField(entry.field)..sort();
        final expected = [...entry.expectedKinds]..sort();
        expect(result, equals(expected));
      });
    }

    test('siempre incluye "any"', () {
      for (final entry in corpus) {
        expect(classifyField(entry.field), contains('any'));
      }
    });

    test('no devuelve duplicados', () {
      for (final entry in corpus) {
        final kinds = classifyField(entry.field);
        expect(kinds.toSet().length, equals(kinds.length));
      }
    });
  });

  group('isFieldCompatibleWithSlot', () {
    SlotSpec mkSlot(List<String> accepts) => SlotSpec(accepts);

    test('slot any-accepts admite cualquier field', () {
      final slot = mkSlot(['any']);
      expect(isFieldCompatibleWithSlot(const FieldSpec('text'), slot), isTrue);
      expect(isFieldCompatibleWithSlot(const FieldSpec('image'), slot), isTrue);
      expect(isFieldCompatibleWithSlot(const FieldSpec('unknown'), slot), isTrue);
    });

    test('slot image admite type=image', () {
      final slot = mkSlot(['image']);
      expect(isFieldCompatibleWithSlot(const FieldSpec('image'), slot), isTrue);
      expect(isFieldCompatibleWithSlot(const FieldSpec('text'), slot), isFalse);
    });

    test('slot text-short NO admite color_hex (KRO-69 exception)', () {
      final slot = mkSlot(['text-short']);
      expect(isFieldCompatibleWithSlot(const FieldSpec('text', behavior: 'color_hex'), slot), isFalse);
    });

    test('slot text-short admite number', () {
      final slot = mkSlot(['text-short']);
      expect(isFieldCompatibleWithSlot(const FieldSpec('number'), slot), isTrue);
    });

    test('slot color SOLO admite behavior=color_hex', () {
      final slot = mkSlot(['color']);
      expect(isFieldCompatibleWithSlot(const FieldSpec('text', behavior: 'color_hex'), slot), isTrue);
      expect(isFieldCompatibleWithSlot(const FieldSpec('text'), slot), isFalse);
      expect(isFieldCompatibleWithSlot(const FieldSpec('image'), slot), isFalse);
    });

    test('slot badge admite rating/enum/ordinal_enum', () {
      final slot = mkSlot(['badge']);
      expect(isFieldCompatibleWithSlot(const FieldSpec('number', behavior: 'rating'), slot), isTrue);
      expect(isFieldCompatibleWithSlot(const FieldSpec('array<string>', behavior: 'enum'), slot), isTrue);
      expect(isFieldCompatibleWithSlot(const FieldSpec('text', behavior: 'ordinal_enum'), slot), isTrue);
      expect(isFieldCompatibleWithSlot(const FieldSpec('text'), slot), isFalse);
    });
  });
}
